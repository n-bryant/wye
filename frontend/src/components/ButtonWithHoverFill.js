import React from "react";
import PropTypes from "prop-types";

import createClassNameHelper from "@n_bryant/classnames-helper";
import JSS_CLASS_NAME_PREFIX from "../../lib/classNamePrefix";
import { withStyles } from "@material-ui/core/styles";

import Icon from "@mdi/react";
import Button from "@material-ui/core/Button";
import Typography from "@material-ui/core/Typography";

import styles from "./ButtonWithHoverFill.styles";

/**
 * renders an Button with hover state for an Icon if provided
 */
export const ButtonWithHoverFill = props => {
  const classnames = ButtonWithHoverFill.classnames(props);
  const { icon, label, displayLabel, handleClick, transformText } = props;
  const [hovered, setHovered] = React.useState(false);

  return (
    <Button
      className={classnames.root({
        transformText
      })}
      onMouseEnter={() => setHovered(true)}
      onMouseLeave={() => setHovered(false)}
      onClick={handleClick}
      aria-label={label}
    >
      {displayLabel && <Typography variant="body1">{label}</Typography>}
      {icon.length && (
        <Icon
          className={classnames.element("icon", {
            hovered,
            withLabel: displayLabel
          })}
          path={icon}
        />
      )}
    </Button>
  );
};
ButtonWithHoverFill.classnames = createClassNameHelper(
  `${JSS_CLASS_NAME_PREFIX}ButtonWithHoverFill`
);
ButtonWithHoverFill.propTypes = {
  // styles to apply
  classes: PropTypes.shape({
    root: PropTypes.string,
    hovered: PropTypes.string,
    icon: PropTypes.string,
    iconWithLabel: PropTypes.string,
    iconHovered: PropTypes.string
  }),
  // the icon to render
  icon: PropTypes.string,
  // the label for the Button
  label: PropTypes.string.isRequired,
  // whether to display the label text on the button
  displayLabel: PropTypes.bool,
  // click handler for the IconButton
  handleClick: PropTypes.func.isRequired,
  // whether to transform the text
  transformText: PropTypes.bool
};
ButtonWithHoverFill.defaultProps = {
  classes: {},
  handleClick: () => {}
};

export default withStyles(styles)(ButtonWithHoverFill);
