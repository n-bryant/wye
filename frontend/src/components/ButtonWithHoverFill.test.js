import { ButtonWithHoverFill } from "./ButtonWithHoverFill";

describe("ButtonWithHoverFill", () => {
  const props = {
    classes: {
      root: "root",
      hovered: "hovered",
      icon: "icon",
      iconWithLabel: "iconWithLabel",
      iconHovered: "iconHovered"
    },
    icon: "icon",
    label: "label",
    displayLabel: true,
    handleClick: jest.fn()
  };

  let wrapper;
  beforeEach(() => {
    const hovered = true;
    const setHovered = jest.fn();
    const useStateSpy = jest.spyOn(React, "useState");
    useStateSpy.mockImplementation(() => [hovered, setHovered]);
    wrapper = shallow(<ButtonWithHoverFill {...props} />);
  });

  it("should render successfully", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should call the passed click handler when clicked", () => {
    wrapper.prop("onClick")();
    expect(props.handleClick).toHaveBeenCalled();
  });
});
