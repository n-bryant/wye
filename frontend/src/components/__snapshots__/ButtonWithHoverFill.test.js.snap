// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonWithHoverFill should render successfully 1`] = `
<WithStyles(ForwardRef(Button))
  aria-label="label"
  className="wye-ButtonWithHoverFill wye-ButtonWithHoverFill--transformText root"
  onClick={[MockFunction]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <WithStyles(ForwardRef(Typography))
    variant="body1"
  >
    label
  </WithStyles(ForwardRef(Typography))>
  <Icon
    className="wye-ButtonWithHoverFill__icon wye-ButtonWithHoverFill__icon--hovered wye-ButtonWithHoverFill__icon--withLabel icon iconHovered iconWithLabel"
    color="currentColor"
    horizontal={false}
    path="icon"
    rotate={0}
    size={null}
    spin={false}
    vertical={false}
  />
</WithStyles(ForwardRef(Button))>
`;
